output/datadog/chunk.go

```diff
@@ -0,0 +1,101 @@
+package datadog
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/relex/gotils/logger"
+	"github.com/relex/slog-agent/base"
+	"github.com/relex/slog-agent/output/shared"
+	"github.com/relex/slog-agent/util"
+)
+
+type intermediateChunk struct {
+	id                   string
+	numRecords, numBytes int
+	maxRecords, maxBytes int
+	compressor           io.WriteCloser // could be something like a gzip.Writer or nil to disable compression
+	writeBuffer          *bytes.Buffer  // an actual buffer that compressor writes to
+}
+
+func buildNewChunkFunc(log logger.Logger, maxRecords, maxBytes int) shared.NewChunkFunc {
+	return func(id string, writeBuffer *bytes.Buffer) shared.Chunker {
+		chunk := &intermediateChunk{
+			id:          id,
+			maxRecords:  maxRecords,
+			maxBytes:    maxBytes,
+			compressor:  nil,
+			writeBuffer: writeBuffer,
+		}
+
+		chunk.compressor = shared.InitGzipCompessor(log, chunk.writeBuffer)
+
+		_, err := chunk.compressor.Write([]byte("["))
+		if err != nil {
+			log.Error(err)
+		} else {
+			chunk.numBytes += len("[")
+		}
+
+		return chunk
+	}
+}
+
+// Write appends new log to log chunk
+func (chunk *intermediateChunk) Write(data base.LogStream) error {
+	var err error
+
+	if chunk.numRecords != 0 {
+		data = append(base.LogStream(","), data...)
```

Comment:
This is a very costly operation for large logs, though this is still POC and not optimized for production.
A TODO comment here would help.

Reply:
BTW it's also the reason we use rewriters for some of the transforms, e.g. [inlining fields](https://github.com/relex/slog-agent/blob/master/rewrite/rinline/rinline_test.go), during the last step before serialization to reduce repeated copying of the message field.

This mainly affects error dumps where each log could be like 30-100KB, and we still want to keep like 1GB/s throughput on production servers for each pipeline.

Reply:
Yea, I guess at least writing directly to the buffer/compressor rather than building a new slice would be faster, fixed